### PR TITLE
detools: Binary delta encoding in C

### PR DESCRIPTION
**Pull request by eerimoq**
Add version 0.51.0.

Based on http://www.daemonology.net/bsdiff/ and HDiffPatch, with the following features:

* bsdiff algorithm.

* sequential or in-place (resumable) patch types.

* LZ4, LZMA, heatshrink or CRLE compression.

* Sequential patches allow streaming.

* Incremental apply patch, suitable for memory constrained embedded
  devices. Only the sequential patch type is supported.

* SA-IS or divsufsort instead of qsufsort for bsdiff.

### CODE REVIEW

**Review by intelligide — CHANGES_REQUESTED** [review]

**Review by uilianries — COMMENTED** [review]
It looks good, only need to shape the package ID

File: recipes/detools/all/test_package/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        old_path = os.path.join(self.source_folder, "old")
+        patch_path = os.path.join(self.source_folder, "patch")
+        new_path = os.path.join(self.source_folder, "new")
+        patched_path = os.path.join(self.build_folder, "patched")
+
+        self.run(f"{bin_path} {old_path} {patch_path} {patched_path}",
+                 run_environment=True)
+
+        if not filecmp.cmp(patched_path, new_path):
+            raise Exception("Patched file does not match new")
```
[the remark this answers is not in the document]
Comment (eerimoq) [reply recipes/detools/all/test_package/conanfile.py:37]: Why should the check be removed?

[the remark this answers is not in the document]
Comment (intelligide) [reply recipes/detools/all/test_package/conanfile.py:37]: the `test_package` is just there to validate that the package is correctly created, and that package consumers will be able to link against it. This is not meant to test if the library is working properly. It needs to be able to be built and run without link errors.

**Review by intelligide — DISMISSED** [review]

**Review by SSE4 — DISMISSED** [review]

File: recipes/detools/all/test_package/test_package.c [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+    if (argc != 4) {
+        printf("Wrong number of arguments.\n");
+
+        return EXIT_FAILURE;
+    }
+
+    res = detools_apply_patch_filenames(argv[1], argv[2], argv[3]);
+
+    if (res == 2780) {
```
Comment (SSE4) [inline recipes/detools/all/test_package/test_package.c:16]: what is 2780? some constant, or comment could help to understand over the magic number

**Review by uilianries — APPROVED** [review]

**Review by intelligide — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Pull request merged**